File: apps/cli/lib/telemetry.ts
import { homedir } from "os";
import { join } from "path";
import {
  readFileSync,
  writeFileSync,
  existsSync,
  mkdirSync,
} from "fs";
import { randomUUID } from "crypto";
import { PostHog } from "posthog-node";
import packageJson from "../package.json" with { type: "json" };

const POSTHOG_API_KEY =
  process.env["POSTHOG_API_KEY"] ?? "phc_REPLACE_WITH_YOUR_KEY";
const POSTHOG_HOST =
  process.env["POSTHOG_HOST"] ?? "https://eu.i.posthog.com";

export const CLI_VERSION = packageJson.version;

const CONFIG_DIR = join(homedir(), ".flowershow");
const CONFIG_FILE = join(CONFIG_DIR, "config.json");

interface CliConfig {
  distinctId?: string;
  telemetryNoticeShown?: boolean;
}

function readConfig(): CliConfig {
  if (!existsSync(CONFIG_FILE)) return {};
  try {
    return JSON.parse(readFileSync(CONFIG_FILE, "utf-8")) as CliConfig;
  } catch {
    return {};
  }
}

function writeConfig(config: CliConfig): void {
  if (!existsSync(CONFIG_DIR)) {
    mkdirSync(CONFIG_DIR, { recursive: true });
  }
  writeFileSync(CONFIG_FILE, JSON.stringify(config, null, 2), "utf-8");
}

function getOrCreateDistinctId(): string {
  const config = readConfig();
  if (config.distinctId) return config.distinctId;
  const id = randomUUID();
  writeConfig({ ...config, distinctId: id });
  return id;
}

function showNoticeIfNeeded(): void {
  const config = readConfig();
  if (config.telemetryNoticeShown) return;
  console.log(
    "\nTelemetry Notice: FlowerShow CLI collects anonymous usage data to improve the product.",
  );
  console.log(
    "To opt out, set the FLOWERSHOW_TELEMETRY_DISABLED=1 environment variable.\n",
  );
  writeConfig({ ...config, telemetryNoticeShown: true });
}

export function isEnabled(): boolean {
  return !process.env["FLOWERSHOW_TELEMETRY_DISABLED"];
}

let _client: PostHog | null = null;

function getClient(): PostHog | null {
  if (!isEnabled()) return null;
  if (!_client) {
    _client = new PostHog(POSTHOG_API_KEY, {
      host: POSTHOG_HOST,
      flushAt: 1,
      flushInterval: 0,
    });
    showNoticeIfNeeded();
  }
  return _client;
}

export function capture(
  event: string,
  properties: Record<string, unknown>,
): void {
  const client = getClient();
  if (!client) return;
  const distinctId = getOrCreateDistinctId();
  client.capture({ distinctId, event, properties });
}

export async function flushTelemetry(): Promise<void> {
  if (!_client) return;
  await _client.shutdown();
}

File: apps/cli/internal/telemetry/telemetry.go
package telemetry

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/google/uuid"
	"github.com/posthog/posthog-go"
	"github.com/flowershow/publish/internal/config"
)

type cliConfig struct {
	DistinctID          string `json:"distinctId,omitempty"`
	TelemetryNoticeShown bool  `json:"telemetryNoticeShown,omitempty"`
}

func configDir() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".flowershow")
}

func configFile() string {
	return filepath.Join(configDir(), "config.json")
}

func readConfig() cliConfig {
	b, err := os.ReadFile(configFile())
	if err != nil {
		return cliConfig{}
	}
	var cfg cliConfig
	json.Unmarshal(b, &cfg)
	return cfg
}

func writeConfig(cfg cliConfig) {
	dir := configDir()
	os.MkdirAll(dir, 0700)
	b, _ := json.MarshalIndent(cfg, "", "  ")
	os.WriteFile(configFile(), b, 0600)
}

func getOrCreateDistinctID() string {
	cfg := readConfig()
	if cfg.DistinctID != "" {
		return cfg.DistinctID
	}
	id := uuid.NewString()
	cfg.DistinctID = id
	writeConfig(cfg)
	return id
}

func showNoticeIfNeeded() {
	cfg := readConfig()
	if cfg.TelemetryNoticeShown {
		return
	}
	fmt.Println("\nTelemetry Notice: Flowershow CLI collects anonymous usage data to improve the product.")
	fmt.Println("To opt out, set the FLOWERSHOW_TELEMETRY_DISABLED=1 environment variable.\n")
	cfg.TelemetryNoticeShown = true
	writeConfig(cfg)
}

// IsEnabled reports whether telemetry is enabled.
func IsEnabled() bool {
	return os.Getenv("FLOWERSHOW_TELEMETRY_DISABLED") == ""
}

var client posthog.Client

func getClient() posthog.Client {
	if !IsEnabled() {
		return nil
	}
	if client == nil {
		var err error
		client, err = posthog.NewWithConfig(config.PostHogAPIKey(), posthog.Config{
			Endpoint: config.PostHogHost(),
		})
		if err != nil {
			return nil
		}
		showNoticeIfNeeded()
	}
	return client
}

// Capture sends a telemetry event.
func Capture(event string, properties map[string]interface{}) {
	c := getClient()
	if c == nil {
		return
	}
	distinctID := getOrCreateDistinctID()
	props := posthog.NewProperties()
	for k, v := range properties {
		props.Set(k, v)
	}
	c.Enqueue(posthog.Capture{
		DistinctId: distinctID,
		Event:      event,
		Properties: props,
	})
}

// Flush flushes and shuts down the telemetry client.
func Flush() {
	if client != nil {
		client.Close()
		client = nil
	}
}
